### DISCUSSION TITLE
Exploring Nanochat

### DISCUSSION THREAD

**Discussion by chrisjmccormick** [KellerJordan/modded-nanogpt#206] [Show and tell]
I've been having a lot of fun this past week exploring Karpathy's nanochat--it's a beautiful project. 

Here's a handful of high level things I find cool about it so far, and then I've got some more specific insights to share comparing our code bases.

### **Vocabulary**

The `speedrun.sh` script runs the full end-to-end process, including training a tokenizer from scratch as step 1. 

I figured this would be a time consuming step that I'd want to split off / skip over, but it's actually quite fast. Maybe a couple minutes? It learns a 64K token vocabulary from 2B tokens.

### **Scaling Laws**

One of the coolest features, I think, is that it's full of integrated scaling laws. You can pick a number of layers for the model, and then architecture parameters, design patterns, dataset size, even some things around learning rates, are derived from that.

### **Evaluation**

It has wandb logging nicely integrated, and includes some benchmark tasks which are automatically run (periodically as well as at the end?).

### **Generation**

This is one of the coolest parts to me. For multiple reasons (one being that the outcome is a chatbot) it includes an efficient engine for generating text. Besides that being useful, I like that it forces some consideration around the impact of our architecture choices on text generation. For example, I think it motivates the use of GQA, which doesn't help us for speedrun pre-training, but is beneficial for efficient generation.

### **RL**

This is the part that I'm actually the most interested in. I've been studying / playing with RL recently. It's really cool what you can accomplish with RL, but it's also obscenely slow. (I think it's the text generation that's primarily to blame?) 

Maybe I'll discover otherwise, but RL seems ripe for speedrunning. With `modded`, we get to throw out the rules, throw out concerns over compatibility and ease of use, and just focus on figuring out what works. That process has lead to a number of innovations in `modded` which apparently translate well enough to larger scales and general use that they're being absorbed into nanochat. (Which, like nanoGPT, I assume will become a gold standard implementation that many papers and projects are built from!).

I'm hoping to find a good target--akin to "3.28 validation loss on fineweb"--that could become an interesting speedrun objective.

### **Challenging the Scaling Laws**

The other thing I find exciting, from a speedrun perspective, is that the built in scaling laws + benchmarking support give us great targets. 

You can configure nanochat for, e.g., 16 layers, get all of the scores for that configuration, and then hitting those metrics faster becomes our target. 

Beyond being efficiently implemented, `modded` is also amazing for how little text it requires to hit its target. I'm looking forward to trying out our various innovations in that framework to see just how much we can break chinchilla. :) 

### **modded vs. nanochat**

I pulled some trace files running both modded and nanochat on an 8xA100 instance and poured through them--I'll start writing up individual details in this thread.

**Comment by chrisjmccormick** [comment]
# **Overall Performance**

I unfortunately couldn't get my hands on an 8xH100 when I was doing all of this, so I was on an 8xA100 for these traces, which means no FP8 head and no fused-squared-relu kernel.

Still plenty of interesting things to look at, though.

### **Fair Comparison**


_Layers_

nanochat let's me specify 11 layers, so that's what I ran to compare against our small track.

_Batch Size_

Our batch sizes are different--nanochat's is 512K tokens, and ours starts at 128K. (Karpathy notes in the code that he opted against the batch size schedule due to complexity). I adjusted nanochat's batch size down to 128K so I could compare some early training steps.

_FlashAttention on A100_

Currently, nanochat kinda barfs if you can't run FA3, which requires an H100--it falls back on a very slow implementation. 

FA2 doesn't have the kv_cache kernel that the generation code relies on, so that makes sense. But for training, FA2 _does_ support the var_len kernel we use, so I tweaked the code to use FA2 instead of the fallback. I may do a pull request for that feature if I can find a good way to automate the FA2 "install". 

_Vocabulary Size_

The vocabulary sizes are different, and I did adjust for that, but first it's pretty interesting to see the impact that a 64K vocab has on things vs. 50K.

### **Vocabulary Size**

Tokenizers and vocab size are something I'm still gaining a deeper understanding of. Smaller vocabs are more efficient to train per-token, but can require more tokens to cover the same data / generate the same reply to a prompt.

For small model sizes, the LM head is the single most computationally brutal step in the entire training process:

<img width="2957" height="949" alt="image" src="https://github.com/user-attachments/assets/694699d6-3ac9-460d-9e52-8f4441ae3321" />

(Note: Steps vary, so these numbers are rough. The first comm in particular can vary, presumably from syncing the 8 gpus)

Given the cost of the vocabulary, and what I know below about GPT-2's tokenizer vs. fineweb, my instinct has always been "why isn't the ideal vocab more like 16k?"

**GPT-2 Tokenizer on Fineweb**

The below is a tangent--more specific to modded than nanochat, since nanochat learns a custom vocab from its training data, but it's fun to share. 

The GPT-2 token frequency statistics on fineweb are insane.

<img width="2126" height="855" alt="image" src="https://github.com/user-attachments/assets/a445956c-d162-41c6-a3ff-95842411c4ca" />


If you tally up the token occurrences in the first 900M tokens of fineweb, you only need the top 9,483 tokens in the GPT-2 vocabulary to represent 99.999% of the data.

Or put another way, the top 0.001% get to consume ~40% of the resources. Hehe.

**nanochat vocab performance**

nanochat reports compression ratios for its tokenizer on its training data, so I'm curious to look at those more closely, and maybe do this same token-frequency plot.

After capturing the above trace at 64K, I did another at the rounded GPT-2 vocab size (50,304) and the time spent on the lm_head appears to scale linearly.

The time drops from 25.4ms --> 19.6ms, and

$ \frac{19.6}{25.4} \times \text{64K} =  50,571$

It will be cool to get to play with this vocab size balance--something we don't get to touch in modded.


### **Overall Step Comparison**

I'll have to redo this once I get an opportunity to run on an 8xH100 again. Because it's on 8xA100, it doesn't include the benefits of the FP8 head and the MLP ReLU^2 kernel.

<img width="5356" height="1495" alt="image" src="https://github.com/user-attachments/assets/fa72f3a1-a280-4e80-a2d7-08e3cc4403b5" />

Some quick observations:

* nanogpt has heavy comms because its VE pattern is every-other-layer + always the final layer (i.e., lots of VE tables with no re-use!)
* The final big matmul in modded's backward belongs to the LM head.
* Our backward pass looks very different due to the parameter banks: 
    * The long stretch of teal kernels at the end--these are normally interleaved into the ops, which is why our backwards layers look so much shorter.
    * These are a symptom of the 'select backwards' kernels that I tried to avoid. They're not inherently bad, but there's likely some performance hits there--hopefully we can resolve it!

When looking at the trace files, I like to use flash attention as the anchor points for layers (because they have a clear kernel name and a distinct color), so that's how the below illustration is aligned.

It starts with the flash attention kernel from layer 1, and then ends around the flash attention kernel for layer 2.

<img width="4178" height="1188" alt="image" src="https://github.com/user-attachments/assets/2e853397-4ca7-4c93-b0fb-dbc0ea550284" />

* nanochat uses standard flash attention; varlen is twice as fast here.
* Not fusing QKV is moderately slower.
* We have an attention gate, nanochat does not.
* Would have been great to see the relu-squared kernel here!

**Comment by andrewbriand** [comment]
> but it's also obscenely slow.

Did you happen to take an approximate breakdown of how long pretraining vs. mid training vs. SFT etc. takes? An RL speedrun would be really cool!

> nanochat let's me specify 11 layers, so that's what I ran to compare against our small track.

Curious, how coherent is the output of the 11-layer model? I was very pleasantly surprised at how the 34 layer model at https://nanochat.karpathy.ai/ can consistently generate grammatical and on-topic text.

**Reply by chrisjmccormick** [reply]
Yup! For the 11-layer model, on 4xH100s:
* The whole thing was ~3 hours.
* Pre-training was 89.11m
* The middle section is a mix of short-ish training segments and evaluation tasks, about 25 minutes.
* RL training was about 50 minutes. 

And some generations:
```
<|bos|>The capital of France is Paris. The city is the capital of France and the largest city in France.
<|bos|>The chemical symbol of gold is Au. It is a silvery-white metal that is very hard and soft. It
<|bos|>If yesterday was Friday, then tomorrow will be Saturday. The weather is going to be hot and humid. The sun is going
<|bos|>The opposite of hot is cold. It is the opposite of cold. It is the opposite of cold.
<|bos|>The planets of the solar system are: Mercury, Venus, Earth, Mars, Jupiter, Saturn, Uranus, Neptune,
<|bos|>My favorite color is red. I love it because it is so cool. I love it because it
<|bos|>If 5*x + 3 = 13, then x is 5*x + 3 = 13. If 5*x
```

It generates an entire report--I'll throw that up here for reference.

And then here are my wandb runs:

https://wandb.ai/chrismccormick/nanochat/runs/0ea3jblv
https://wandb.ai/chrismccormick/nanochat-mid/runs/lygog71b
https://wandb.ai/chrismccormick/nanochat-rl/runs/1odogsdr

It runs a lot of benchmarks:

<img width="3716" height="1924" alt="image" src="https://github.com/user-attachments/assets/6caceb77-e96d-4b94-bfab-cf1f742e543a" />

RL is noisy, confusing, and slow 😁

<img width="2791" height="956" alt="image" src="https://github.com/user-attachments/assets/15429eaf-dbbc-46df-a202-68d04edde223" />

**Reply by chrisjmccormick** [reply]
# nanochat training report

Generated: 2026-01-25 02:51:02

## Environment

### Git Information
- Branch: master
- Commit: 85b3e95 (dirty)
- Message: 320 experiments just to tune the adam beta1 of x0 a little bit up from 0.8 to 0.

### Hardware
- Platform: Linux
- CPUs: 52 cores (104 logical)
- Memory: 885.4 GB
- GPUs: 4x NVIDIA H100 80GB HBM3
- GPU Memory: 316.8 GB total
- CUDA Version: 12.8
- Hourly Rate: $12.00/hour

### Software
- Python: 3.10.12
- PyTorch: 2.9.1+cu128


### Bloat
- Characters: 460,190
- Lines: 10,649
- Files: 48
- Tokens (approx): 115,047
- Dependencies (uv.lock lines): 3,713

Run started: 2026-01-25 02:51:05

---

## Tokenizer training
timestamp: 2026-01-25 02:52:00

- max_chars: 2,000,000,000
- doc_cap: 10,000
- vocab_size: 65,536
- train_time: 51.0079
- num_special_tokens: 9
- token_bytes_min: 1
- token_bytes_max: 32
- token_bytes_mean: 6.9197
- token_bytes_std: 2.8748


## Tokenizer evaluation
timestamp: 2026-01-25 02:52:05

### Comparison with GPT-2

| Text Type | Bytes | GPT-2 Tokens | GPT-2 Ratio | Ours Tokens | Ours Ratio | Relative Diff % |
|-----------|-------|--------------|--------------|-------------|------------|-----------------|
| news | 1819 | 404 | 4.50 | 375 | 4.85 | +7.2% |
| korean | 893 | 745 | 1.20 | 712 | 1.25 | +4.4% |
| code | 1259 | 576 | 2.19 | 492 | 2.56 | +14.6% |
| math | 1834 | 936 | 1.96 | 966 | 1.90 | -3.2% |
| science | 1112 | 260 | 4.28 | 228 | 4.88 | +12.3% |
| fwe-train | 4208518 | 900364 | 4.67 | 856883 | 4.91 | +4.8% |
| fwe-val | 4768657 | 1027270 | 4.64 | 980300 | 4.86 | +4.6% |

### Comparison with GPT-4

| Text Type | Bytes | GPT-4 Tokens | GPT-4 Ratio | Ours Tokens | Ours Ratio | Relative Diff % |
|-----------|-------|--------------|--------------|-------------|------------|-----------------|
| news | 1819 | 387 | 4.70 | 375 | 4.85 | +3.1% |
| korean | 893 | 364 | 2.45 | 712 | 1.25 | -95.6% |
| code | 1259 | 309 | 4.07 | 492 | 2.56 | -59.2% |
| math | 1834 | 832 | 2.20 | 966 | 1.90 | -16.1% |
| science | 1112 | 249 | 4.47 | 228 | 4.88 | +8.4% |
| fwe-train | 4208518 | 874799 | 4.81 | 856883 | 4.91 | +2.0% |
| fwe-val | 4768657 | 1001442 | 4.76 | 980300 | 4.86 | +2.1% |


## Base model training
timestamp: 2026-01-25 04:31:06

- run: d11_baseline
- device_type: 
- depth: 11
- aspect_ratio: 64
- head_dim: 128
- max_seq_len: 2048
- window_pattern: SSSL
- num_iterations: -1
- target_flops: -1.0000
- target_param_data_ratio: 20
- device_batch_size: 32
- total_batch_size: 524,288
- embedding_lr: 0.3000
- unembedding_lr: 0.0040
- weight_decay: 0.2000
- matrix_lr: 0.0200
- scalar_lr: 0.5000
- adam_beta1: 0.8000
- adam_beta2: 0.9500
- warmup_ratio: 0.0000
- warmdown_ratio: 0.4000
- final_lr_frac: 0.0000
- resume_from_step: -1
- eval_every: 250
- eval_tokens: 10,485,760
- core_metric_every: 2000
- core_metric_max_per_task: 500
- sample_every: 2000
- save_every: -1
- model_tag: None
- Number of parameters: 480,511,126
- Number of FLOPs per token: 9.012580e+08
- Calculated number of iterations: 18,330
- Number of training tokens: 9,610,199,040
- Tokens : Params ratio: 20.0000
- DDP world size: 4
- warmup_ratio: 0.0000
- warmdown_ratio: 0.4000
- final_lr_frac: 0.0000
- Minimum validation bpb: 0.8345
- Final validation bpb: 0.8345
- CORE metric estimate: 0.1915
- MFU %: 41.74%
- Total training flops: 8.661269e+18
- Total training time: 89.11m
- Peak memory usage: 33818.89MiB


## Midtraining
timestamp: 2026-01-25 04:39:42

- run: d11_baseline_p2
- device_type: 
- dtype: bfloat16
- model_tag: None
- model_step: None
- num_iterations: -1
- max_seq_len: 2048
- device_batch_size: 32
- total_batch_size: 524,288
- embedding_lr: 0.2000
- unembedding_lr: 0.0040
- matrix_lr: 0.0200
- weight_decay: 0.0000
- init_lr_frac: 1.0000
- eval_every: 150
- eval_tokens: 10,485,760
- dry_run: False
- Number of iterations: 788
- DDP world size: 4
- Minimum validation bpb: 0.4253


## Chat evaluation mid
timestamp: 2026-01-25 04:48:30

- source: mid
- task_name: None
- dtype: bfloat16
- temperature: 0.0000
- max_new_tokens: 512
- num_samples: 1
- top_k: 50
- batch_size: 8
- model_tag: None
- step: None
- max_problems: None
- device_type: 
- ARC-Easy: 0.3636
- ARC-Challenge: 0.2901
- MMLU: 0.3074
- GSM8K: 0.0000
- HumanEval: 0.0671
- SpellingBee: 0.9688
- ChatCORE metric: 0.2195


## Chat SFT
timestamp: 2026-01-25 04:50:10

- run: d11_baseline_p2
- device_type: 
- dtype: bfloat16
- source: mid
- model_tag: None
- model_step: None
- num_epochs: 1
- num_iterations: -1
- device_batch_size: 4
- target_examples_per_step: 32
- embedding_lr: 0.2000
- unembedding_lr: 0.0040
- matrix_lr: 0.0200
- weight_decay: 0.0000
- init_lr_frac: 0.0200
- eval_every: 100
- eval_steps: 100
- eval_metrics_every: 200
- eval_metrics_max_problems: 1024
- Training rows: 22,439
- Number of iterations: 701
- Training loss: 1.4571
- Validation loss: 1.1106


## Chat evaluation sft
timestamp: 2026-01-25 04:57:53

- source: sft
- task_name: None
- dtype: bfloat16
- temperature: 0.0000
- max_new_tokens: 512
- num_samples: 1
- top_k: 50
- batch_size: 8
- model_tag: None
- step: None
- max_problems: None
- device_type: 
- ARC-Easy: 0.3826
- ARC-Challenge: 0.2824
- MMLU: 0.3095
- GSM8K: 0.0470
- HumanEval: 0.0183
- SpellingBee: 0.9922
- ChatCORE metric: 0.2261


## Chat RL
timestamp: 2026-01-25 05:47:41

- run: d11_baseline_p2
- device_type: 
- dtype: bfloat16
- source: sft
- model_tag: None
- model_step: None
- num_epochs: 1
- device_batch_size: 8
- examples_per_step: 16
- num_samples: 16
- max_new_tokens: 256
- temperature: 1.0000
- top_k: 50
- embedding_lr: 0.2000
- unembedding_lr: 0.0040
- matrix_lr: 0.0200
- weight_decay: 0.0000
- init_lr_frac: 0.0500
- eval_every: 60
- eval_examples: 400
- save_every: 60


## Chat evaluation rl
timestamp: 2026-01-25 05:49:35

- source: rl
- task_name: GSM8K
- dtype: bfloat16
- temperature: 0.0000
- max_new_tokens: 512
- num_samples: 1
- top_k: 50
- batch_size: 8
- model_tag: None
- step: None
- max_problems: None
- device_type: 
- GSM8K: 0.0902


## Summary

- Characters: 460,190
- Lines: 10,649
- Files: 48
- Tokens (approx): 115,047
- Dependencies (uv.lock lines): 3,713

| Metric          | BASE     | MID      | SFT      | RL       |
|-----------------|----------|----------|----------|----------|
| ARC-Challenge   | -        | 0.2901   | 0.2824   | -        |
| ARC-Easy        | -        | 0.3636   | 0.3826   | -        |
| GSM8K           | -        | 0.0000   | 0.0470   | 0.0902   |
| HumanEval       | -        | 0.0671   | 0.0183   | -        |
| MMLU            | -        | 0.3074   | 0.3095   | -        |
| ChatCORE        | -        | 0.2195   | 0.2261   | -        |

Total wall clock time: 2h6m

**Reply by andrewbriand** [reply]
Interesting, so pre-training is still the main chunk of time, but RL is also very expensive.

If we did a nanochat track, do you think the data would be as restricted as for the nanogpt track? I'm wondering for example how much the quality of the synthetic data used in midtraining affects the training speed.

**Reply by chrisjmccormick** [reply]
I learned that the settings in the speedrunning script were outdated--it was set to train on 20x the number of model parameters, but apparently only needs 4x. Pre-training nanochat apparently runs on an 8xH100 in more like 5 minutes.

Yeah, data mixture and data generation seem like really interesting problems! I wonder if there's any way to give free reign on that without it polluting the benchmark task. 

Also, I'm realizing that nanochat is already a sort of speed run competition, it's being actively improved, and you can contribute. That gives an added freedom of there not being any specific target other than "make any/all of this better". But it does come with the restriction that stuff be elegant and generalizable. :)
